test/test_serialization.py

```diff
@@ -4128,6 +4148,201 @@ def test_empty_class_serialization(self):
+            restore_methods["__set__"] = None
+
+            try:
+                # Check that weights_only=False load raises the RuntimeErrors we expect
+                with self.assertRaisesRegex(RuntimeError, "running __getattribute__"):
+                    torch.load(f, weights_only=False)
+                subclass.__getattribute__ = restore_methods['__getattribute__']
+                with self.assertRaisesRegex(RuntimeError, "running __setstate__"):
+                    torch.load(f, weights_only=False)
+                subclass.__setstate__ = restore_methods['__setstate__']
+                with self.assertRaisesRegex(RuntimeError, "running __setattr__"):
+                    torch.load(f, weights_only=False)
+                subclass.__setattr__ = restore_methods['__setattr__']
+                # should finally work
+                torch.load(f, weights_only=False)
+
+                # Check that weights_only=True catches that risky methods are overriden
+                subclass.__setstate__ = self._create_bad_func("__setstate__")
+                subclass.__getattribute__ = self._create_bad_func("__getattribute__")
+                subclass.__setattr__ = self._create_bad_func("__setattr__")
+                with self.assertRaisesRegex(pickle.UnpicklingError,
+                                            "methods: __getattribute__=True __getattr__=True __get__=True "
+                                            "__setattr__=True __set__=True __setstate__=True"):
+                    torch.load(f, weights_only=True)
+                risky_methods = ['__get__', '__set__', '__getattr__', '__setattr__', '__getattribute__', '__setstate__']
+                for i, meth in enumerate(risky_methods):
+                    setattr(subclass, meth, restore_methods[meth])
+                    if i != len(risky_methods) - 1:
+                        # When the given methods are not all back to default, load should still throw
+                        # but reflect which methods are no longer overriden
+                        with self.assertRaisesRegex(pickle.UnpicklingError, f"{meth}=False"):
+                            torch.load(f, weights_only=True)
+                    else:
+                        # When the given methods are all back to default, weights_only load should finally work
+                        loaded = torch.load(f, weights_only=True)
+            finally:
+                for method, func in restore_methods.items():
+                    setattr(subclass, method, func)
+                a = subclass(torch.randn(2, 3))
+
+    @skipIfTorchDynamo("name 'SYNTHETIC_LOCAL' is not defined")
+    def test_safe_globals_for_weights_only(self):
+        '''
+        Tests import semantic for tensor subclass and the {add/get/clear}_safe_globals APIs
+        '''
+        # Needed to prevent UnboundLocalError: local variable 'TwoTensor' referenced before assignment
+        global TwoTensor
```

Comment:
Huh why?

Reply:
This is weird to me as well, but seems to happen because TwoTensor is imported again later in the test, can minimize this to this will still throw the error

```python

def test_safe_globals_for_weights_only(self):
    '''
    Tests import semantic for tensor subclass and the {add/get/clear}_safe_globals APIs
    '''
    t = TwoTensor(torch.randn(2, 3), torch.randn(2, 3))
    from torch.testing._internal.two_tensor import TwoTensor
 ```

raises 
```
t = TwoTensor(torch.randn(2, 3), torch.randn(2, 3))
UnboundLocalError: local variable 'TwoTensor' referenced before assignment
```